src/pip/_internal/req/req_install.py

```diff
@@ -219,13 +216,23 @@ def format_debug(self) -> str:
             state=", ".join(state),
         )
 
+    @property
+    def checked_req(self) -> Requirement:
+        assert self.req is not None
+        return self.req
+
```

Comment:
I’m sorry but I don’t really like this approach. If we can’t (via type checks) guarantee that the input is never `None`, then that means we expect the assert to trigger. And we shouldn’t ever be adding asserts that we expect to get triggered - IMO asserts are just for cases that “cannot happen”.

If the problem is that the requirement can be in two states, “under construction” where it’s not safe to assume these values are set up, and “ready”, where everything is definitely not `None`, then we should work out how to express that in the type system - and if we can’t, raise the issue with the typing people.

Reply:
Just to clarify, these are all asserts that "cannot happen". I've only added these asserts on code paths that mypy indicates you'd get e.g. a TypeError or AttributeError if something was actually None. The current PR should preserve all non-exceptional semantics.

I can't quite tell exactly what you're objecting to... If it's just the use of `property` I'd be happy to inline these asserts wherever they get used. Note that this sort of defensive programming / use of self-documenting preconditions is already commonly used in this file, e.g. see the use of `assert self.source_dir` on all the methods that touch it.

> If the problem is that the requirement can be in two states,

It is possible to express this kind of thing in the type system, but a) it would require much larger changes than this one, b) knowing what I know of your feelings about typing, I suspect you wouldn't like it either ;-)

Reply:
Anyway, I'm here because @pradyunsg indicated such a PR would be welcome and because I do think this is an improvement over unsound type checking.

I'd be happy to inline asserts instead of using properties and address uranusjr's comments. But if your hesitation about the approach goes beyond that, feel free to close this PR :-)

Reply:
I'd actually say that tackling a different file than this one would be a better place to start. :)

That said, inline asserts are definitely a better thing to do here instead of these "checked_foo"  properties.

Reply:
AAARGH!!!! I had an almost-complete extended explanation of my position here, and github (or my browser) threw it away when I went to check a detail of the PR. Apologies if this is a bit disjointed, I'm trying to remember the points I made before...

> It is possible to express this kind of thing in the type system, but a) it would require much larger changes than this one, b) knowing what I know of your feelings about typing, I suspect you wouldn't like it either ;-)

lol, thanks for your understanding :-) After looking more closely, I see a bit better what you're trying to do here. I am trying to look at this from the perspective that type checks are a good thing, apologies if I don't manage to completely hide my reservations[^1].

First of all, I'm honestly rather appalled by mypy's current behaviour here. We tell the type checker that `req` is an `Optional[Requirement]`, and yet it still lets us try to access properties of `req` without ensuring it's not `None`? That's basically just not doing its job. I'm sorry, I don't want to hear any "practicality" arguments here, either check the damn types or don't claim to be a type checker...

So yes, I guess that means I do support making this change. My concern is that we do it in a way that doesn't simply ignore the issue that type checking has found. And my worry is that asserts do precisely that - they simply add a runtime check that replicates the type check, and yet doesn't give any more user friendly result than the exception you get if you try to access an attribute of `None`. If we're willing to dump a traceback on the user anyway, why not just use `cast` to ignore the type[^2]?

More realistically, I do think we should consider refactoring `InstallRequirement` so that we can somehow say "until this happens, `req` is optional, and we can't use methods that expect `req` to be set". As things currently stand, I don't even know when that point would be, so that would definitely be a benefit. But that involves a potentially major refactor of one of the ugliest bits of pip, so I'm certainly not suggesting it's something we should tackle lightly.

So where does that leave us? Apparently, mypy knew about a problem in our code, so that's good - sort of. It seems like we have a few options:

1. Defer fixing the issue until we can fix it "properly". This could take some time, and in the meantime we have to allow mypy to continue ignoring the issue.
2. Try to find a way to check and report the problem at runtime, in a way that gives the user proper feedback on what went wrong (that would likely need to be done on a case by case basis).
3. Patch over the problem with a runtime assert, to satisfy mypy but without improving the user experience if something goes wrong.
4. Add complicated type machinery to express what's happening properly. I'm not *against* this, despite your suspicions, but I am worried that the more we use complex type features, the more we add to the maintenance burden as few of the maintainers will have the skills to work with that code.
5. Do something adhoc like `getattr(self.req, "name", "<unknown requirement>")`. Is that better than an error? Who knows?
6. Do nothing. Accept that we can't type this properly, but we don't want to spend time fixing the underlying problem.

At the moment, I'm not sure this problem is important enough to need anything more than (6). But if it does, I think we should at least consider a better solution than just adding asserts (which let us fix the types, but otherwise gain us nothing).

@pypa/pip-committers what do you think? Am I being too strict[^3] here?

> I'd be happy to inline asserts instead of using properties and address uranusjr's comments. But if your hesitation about the approach goes beyond that, feel free to close this PR :-)

If we go with asserts, I agree that a property is a good way of handling it. And I certainly don't intend to just close this PR - your work is appreciated even if I have questions about the approach. Ultimately, all I wanted to do here was to raise the question of whether we should do something to tackle the underlying issue, and to make the point that if we're going to implement type checking, we should take type errors seriously - and at least consider other options before asserting.

[^1]: I'm slowly coming around to understanding the benefits of type checking, but I'm not all the way there yet...
[^2]: Don't take that too seriously, I know cast isn't meant to be used like that...
[^3]: Pun intended 🙂

Reply:
> Patch over the problem with a runtime assert, to satisfy mypy but without improving the user experience if something goes wrong.

IMO, the problem here is of bad design of this object -- which we've discussed at length in #6607. :)

Reply:
So you're saying that we take the type error as flagging an implied contract that we've not been enforcing, rather than an actual error? OK, I can see that. And if that's the way we're looking at it, I agree that asserting at the point of use is more appropriate than having a `checked_*` property.

There's still something that makes me uncomfortable about this approach to type errors. I'm finding it very hard to put my reservation into words, but essentially it feels like we're not actually fixing type errors, but rather just finding ways to shut the checks up. Maybe that's just an artifact of the process of adding type annotations to an established codebase. But if we establish a "the code is right, get the type checker to understand that" mindset, we're not getting the benefit from type annotations.

🤷 At this point, I've probably spent more time (of mine and everyone else's) on this than it could ever justify. So I'll leave it there.

Reply:
I definitely agree that replacing static type checks with runtime type checks alone doesn't get anyone very far. Static analysis works best when it confirms your invariants for you, which is not what happens with assert.

In fact, my first attempt at this PR was to disallow InstallRequirement from taking req=None, but I lack familiarity with the codebase and it ended up becoming a daunting change.

I do think there is still some incremental benefit to option 3 over your option 1:
1) It lets us get rid of `# mypy: strict-optional=False`, which I agree is appalling. Doing so allows the type checker to point out potential problems when modifying or refactoring, even if we paper over existing issues to some extent.
2) It states our invariants more clearly. Right now one has to trace code to know whether a method can be called when self.req is None.
3) The user experience is marginally better with an assert, as opposed to e.g. "None has no attribute" on some line somewhere and figuring out how far back in the stack that None comes from.

I'll go ahead and inline the checked* properties. I agree that an inline assert more clearly documents implied contracts.

Reply:
Thanks for your patience while I thought this through in public 🙂